cpp/dolfinx/refinement/interval.h

```diff
@@ -0,0 +1,225 @@
+// Copyright (C) 2024 Paul Kühner
+//
+// This file is part of DOLFINX (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <mpi.h>
+
+#include <concepts>
+#include <cstdint>
+#include <optional>
+#include <stdexcept>
+#include <vector>
+
+#include "dolfinx/mesh/Mesh.h"
+#include "dolfinx/mesh/cell_types.h"
+#include "dolfinx/mesh/utils.h"
+#include "dolfinx/refinement/plaza.h"
+
+namespace dolfinx::refinement
+{
+
+namespace impl
+{
+
+/// Refine with markers returning new mesh data.
+///
+/// @param[in] mesh Input mesh to be refined
+/// @param[in] cells Indices of the cells that are marked for refinement
+///
+/// @return New mesh data: cell topology, vertex coordinates and parent
+/// cell indices.
+template <std::floating_point T>
+std::tuple<graph::AdjacencyList<std::int64_t>, std::vector<T>,
+           std::array<std::size_t, 2>, std::vector<std::int32_t>>
+compute_interval_refinement(const mesh::Mesh<T>& mesh,
+                            std::optional<std::span<const std::int32_t>> cells)
+{
+  auto topology = mesh.topology();
+  assert(topology);
+  assert(topology->dim() == 1);
+
+  auto map_c = topology->index_map(1);
+  assert(map_c);
+
+  // TODO: creation of sharing ranks in external function? Also same code in use
+  // for plaza
+  // Get sharing ranks for each cell
+  graph::AdjacencyList<int> cell_ranks = map_c->index_to_dest_ranks();
+
+  // Create unique list of ranks that share cells (owners of ghosts plus
+  // ranks that ghost owned indices)
+  std::vector<int> ranks = cell_ranks.array();
+  std::ranges::sort(ranks);
+  auto to_remove = std::ranges::unique(ranks);
+  ranks.erase(to_remove.begin(), to_remove.end());
+
+  // Convert cell_ranks from global rank to to neighbourhood ranks
+  std::ranges::transform(cell_ranks.array(), cell_ranks.array().begin(),
+                         [&ranks](auto r)
+                         {
+                           auto it = std::lower_bound(ranks.begin(),
+                                                      ranks.end(), r);
+                           assert(it != ranks.end() and *it == r);
+                           return std::distance(ranks.begin(), it);
+                         });
+
+  // create refinement flag for cells
+  // TODO: vector of bools? -> make of use std specialization for type bool
```

Comment:
This is best avoided - specialisation of vector for type bool is considered a mistake in the C++ standard.